### ISSUE TITLE
Rebuild rate limit and gap tables for gauge measurement

### ISSUE THREAD

**Issue by brendam** [digitalmethodsinitiative/dmi-tcat#168]
On new servers with the recent dmi-tcat version I'm getting 1000's of emails warning that a rate limit is reached. Overnight a new tracker sent over 14,000 emails and has had it's ip address blocked by my email provider (where the mail was being sent to).

config parameters were left at default values:

```
define('IDLETIME', 600);
define('RATELIMIT_SILENCE', 300);
define('RATELIMIT_MAIL_HOURS', 24);
```

I've changed RATELIMIT_MAIL_HOURS to 0 to stop sending emails.

**Comment by ErikBorra** [comment]
Hi @brendam,

that's very inconvenient! Could you let me know the sha of the latest commit of a server which has the issue, as well as one for a server which doesn't have the issue?

Best,

Erik

**Comment by brendam** [comment]
I'm not sure what the most recent ones are that don't have the issue - it seems to be a bit tricky to trigger. Happens when a new server is rate limited, but seemed like it *might* go away if the server was restarted. Was easily repeatable by setting a track on 'the' for any new server. The version on the server from last night has:

```git rev-parse HEAD
6fa087d64bdd298555491cfa8b716fabe422ad20```

and 

```
git status
On branch master
Your branch is behind 'origin/master' by 34 commits, and can be fast-forwarded.
  (use "git pull" to update your local branch)
```
Not sure if that's the right way to check?

**Comment by brendam** [comment]
I'm working on a fix for this - still need to test it - https://github.com/brendam/dmi-tcat/tree/issue-168

**Comment by brendam** [comment]
I haven't used php for a long time and It seems that i don't understand how the rate limiting code is working, my changes don't fix the problem and introduce some different strange things, or maybe I've just noticed them - the start time and end times written to the database only seems to vary in seconds not date or hours and minutes. I don't have more time to spend on it for now, sorry. 

I had some php errors in the first one I posted that I've corrected - here is the latest version: https://github.com/digitalmethodsinitiative/dmi-tcat/compare/master...brendam:issue-168

I think a better approach to stopping too many emails might be to add a boolean field to the tcat_error_ratelimit table and update it for lines that have been reported, then only select ones that haven't been reported - that keeps the testing for that in the mail sending function.  But it still leaves the problem with the times being written to the database.

**Comment by dentoir** [comment]
Latest commit should fix the timestamp problem, but not yet the too many e-mails.

**Comment by dentoir** [comment]
Hi @brendam 

I believe we have fixed the problem in temporary branch issue168. The approach is to create a separate table tcat_emails which stores type and timestamps of e-mails sent by TCAT.

https://github.com/digitalmethodsinitiative/dmi-tcat/compare/issue168?expand=1

This has the benefit it will not keep sending e-mails whenever the controller restarts a task. It can keep track of other type of e-mails we may want to send in the future as well.
The table is created on-the-fly which is the default for other tcat_* tables.

I've tested this code on a single server which repeatedly reaches a rate limit, and it can only send me a single e-mail. What do you both think?

**Comment by brendam** [comment]
Thanks @dentoir,

Testing your branch now, first email has been sent and only received a single email, will wait hour to see if I get next email. Restarting didn't result in a second email, so that's working too.

Still seems to be writing to the tcat_error_ratelimit table too often (continuously). Here are 4 records:

```
599	track	2016-03-29 20:52:58	2016-03-29 20:52:58	3262
600	track	2016-03-29 20:52:58	2016-03-29 20:52:59	3311
601	track	2016-03-29 20:53:00	2016-03-29 20:53:00	3381
602	track	2016-03-29 20:53:01	2016-03-29 20:53:01	3425
```

My config has:

```
define('RATELIMIT_MAIL_HOURS', 1);
define('RATELIMIT_SILENCE', 300);
```

I don't know if it maters, but tcat_error_ratelimit is in Brisbane time and the tcat_emails is in GMT 
```
1	ratelimit	2016-03-29 10:44:03
```

**Comment by brendam** [comment]
The second email was sent after the 1 hour delay. I've got over 6,000 entries in the `tcat_errror_ratelimit` table. Turned off collecting again for now.

**Comment by dentoir** [comment]
The problem was that the ratelimit tracking was designed with 'bursts' of rate limit problems in mind and therefore it tracked whenever a rate limit situation 'started' and 'stopped'. However, this model leads to too much registrations and complexity, especially on heavy load servers. It makes more sense to simply register periodically the ammount of tweets that where rate limited. My latest commit in the issue168 branch does this exactly this and it simplified things a lot. The granularity is defined at 1 hour; the tcat_error_ratelimit table structure is not modified, but start and end is now based on hours.

The RATELIMIT_SILENCE parameter is no longer used and is removed from the config template.

If everything works, this is a good step forward to implementing issue #150 as well.

**Comment by brendam** [comment]
I agree, the new code looks much cleaner. But I tested your commit from yesterday, c5529a72a222f4824925f155d68775ad5d211a55, this morning and just tested the latest commit and neither seem to ever detect rate limiting. Is it working for you?

I'm tracking 'the,and' which should hit rate limiting.

**Comment by dentoir** [comment]
Yes, it is working, but the current code is recording only whole hours (see documentation in an  even more recent commit). So it will start recording at minute 0 if a new hour, and not immediatly.

We've discussed this behaviour this morning and we want to increase the granularity to minutes instead of hours. This will mean 3600 records per day, which seems acceptable. That changes will be made first in this branch.

**Comment by brendam** [comment]
Ah, ok. I wasn't patient enough. I'll try again.

**Comment by brendam** [comment]
Leaving it for longer worked, thanks.

**Comment by dentoir** [comment]
I've pushed a new change to the issue168 branch, which should change the recording interval to one minute. The code did become a bit more complex again, and the current code is verbose in logging. If you've used this branch and want a historical record, its probably best to remove old (hourly) records you captured in this branch, because the numbers where not accurate yet.

The increased complexity is due to the fact that we want to measure every single minute and zero the minutes where there was no capturing at all (because you may have a tiny queryset) back in time. This would make producing charts of rate limits easier.

**Comment by brendam** [comment]
There is debug information being continuously written to the track.error.log

```
2016-04-07 12:10:13	(debug) at measurement minute 10, we are reading 18423 tweets rate limited (record grows to 965909)
2016-04-07 12:10:14	(debug) at measurement minute 10, we are reading 18236 tweets rate limited (record grows to 984145)
q2016-04-07 12:10:14	(debug) at measurement minute 10, we are reading 18511 tweets rate limited (record grows to 1002656)
2016-04-07 12:10:14	(debug) at measurement minute 10, we are reading 18279 tweets rate limited (record grows to 1020935)
2016-04-07 12:10:14	(debug) at measurement minute 10, we are reading 18269 tweets rate limited (record grows to 1039204)
2016-04-07 12:10:14	(debug) at measurement minute 10, we are reading 18352 tweets rate limited (record grows to 1057556)
2016-04-07 12:10:15	(debug) at measurement minute 10, we are reading 18460 tweets rate limited (record grows to 1076016)
```
records are being written once per minute to the `tcat_error_ratelimit` table.

email was sent at start of event.  I'll wait for an hour to check that the next email gets sent (I've set `define('RATELIMIT_MAIL_HOURS', 1);`

**Comment by brendam** [comment]
The 2nd email was sent as expected after 1 hour. So this seems to be working now, just need to turn off some of the debug messages.

**Comment by dentoir** [comment]
Okay, to summarize the current status of this issue, which is unfortunately not ready yet to merge into master.

* This issue168 branch does a gauge per minute, which is the desired future behaviour.
* The current master branch keeps an incremental counter per minute, which is the deprecated, but stable behaviour.
* The **old** ratelimit table entries did (up until very recent ago) keep an incremental counter per hour, because of a bug you encountered in the datetime function.

To futher complicate things a bit:

* The master branch does not produce entries in the _ratelimit table for periods where there is no rate limit (zeros). We want 'no ratelimits' to be zero.

If we don't care about the historic data, we can just truncate and be on our way with this new branch, but otherwise, we need a script to re-assemble the _ratelimit table, with as much precision as possible. I've been working on this and I've got a functioning re-assembly script; I just want to do one more big test. The code has been commited to this branch (common/upgrade.php !)

Regarding merging everything to master, there is still one obstacle:

* If people checkout the new code without *immediatly* running a common/upgrade.php - we will have some point-in-time where the incremental counter becomes a gauge (because this is done in the capture script and effective immediatly) and the re-assemble script obviously needs to know this / work only on the older data.

So we need to keep a timestamp somewhere, and handle this.

If anyone working on this branch and wants to test the upgrade.php script, remove the entries from tcat_error_ratelimit up until the point you switched to this branch. The upgrade.php script does backup the complete ratelimit table under analysis/cache before destructively rebuilding it.

I'd appreciate anyone testing the current code. Though this stuff is a bit difficult to visualize without charts or export modules (issue #160 again).

**Comment by dentoir** [comment]
Hi,

Latest code relating to this issue is here: https://github.com/digitalmethodsinitiative/dmi-tcat/tree/issue168gap

It will get pushed into master soon. The upgrade code assumes a state where the gauge-per-minute is not yet activated, so if you're running a server with the issue168 branch @brendam , you'd need to change a line in the upgrade.php script to tell it to only work on the older timeframe.

Afterwards analysis scripts should be forthcoming as well.

**Comment by ErikBorra** [comment]
Awesome work @dentoir!

**Comment by brendam** [comment]
@dentoir I'm not running a production server with the issue branch - was only doing testing. Sorry I've been slow to respond, been busy with other things. Thanks for fixing the emails problem!

**Comment by brendam** [comment]
@dentoir I just did a fresh install using the new command line install script, installed sendmail and then checked out the issue168gap branch and ran the update.php script in common. 

Using track of 'a,b,c' I got rate limited and was sent a lot of emails (130 in 2 minutes). I had `define('RATELIMIT_MAIL_HOURS', 1);` because I wanted to see if I got the 2nd email 1 hour later.

Did I miss something about how i should be configuring the branch?

**Comment by dentoir** [comment]
Hi @brendam 

This is really strange, I can't see anything wrong with the code at first glance. I'll be sure to take a look at this next Monday or Tuesday. We want to be able to merge this code into master, and this problem should've disappeared them. In the mean time, I neglected to mention the latest branch is now 'phrasestable', which improves the ability to export rate limit information. It will not solve the e-mail problem, however, because that code has not changed.

Unfortunately I can't look at this today. If you care to help debugging the code is in common/functions.php - if I were to debug this, I'd run dmitcat_track.php directly from the command-line and change the mail() function (obviously!) into a print statement. I'd want to dump what you receive from the SQL statement (print_r($row)). It should return 0 (so as to not send multiple e-mails), but apparently it does not work. Sorry for the spammy behaviour of the current code. It seems to be triggered in a heavily rate limited query, but I haven't reproduced it myself (yet).

Cheers and keep you posted,

Emile

**Comment by brendam** [comment]
I'll try out the phrasestable branch and see what happens.

**Comment by brendam** [comment]
Sorry, I think I made a mistake in checking out the remote branch and done my testing on master instead - not understanding the change to having only the master fetched by default now. 

Just sorted that out by changing the git config and checked out origin `issue168gap` branch instead of making a new local branch by that name. Testing that and it looks like it is only ending one email per hour as intended. I'll leave it running and see if I get a second email.

**Comment by brendam** [comment]
I got the second email, and I've checked the database and it looks good too. Setting the email address to  `ubuntu@localhost` for testing avoids any problems with a lot of emails.

As far as I can tell this is all working now.

**Comment by ErikBorra** [comment]
Code review was just sent out, so we should be able to merge this into master soon!

**Comment by brendam** [comment]
Do you know when this can be merged? It's getting out of step with master.

**Comment by dentoir** [comment]
Hi @brendam,

The code is ready to be merged, but unfortunately it is blocked by this issue (#197), for which a fix will need to be merged first. We are rebuilding gap and ratelimits based on historicial data, but that historical data should be in the accurate timezone (which is especially relevant for gap detection). We are working on this issue the current week.

**Comment by dentoir** [comment]
Everything from master has been already pulled into the phrasestable branch by-the-way, so the final merge itself shouldn't be difficult. We'll work on the phrasestable branch as well to fix issue #197

**Comment by brendam** [comment]
thanks, that sounds good.

**Comment by dentoir** [comment]
Fixed and merged in master! Please see my last comment on issue #197

**Comment by brendam** [comment]
Installing the latest version using the Australian Launchpod deploy service (https://launchpod.intersect.org.au) which I am testing, the TCAT pages fail to load after install. Looking into it, the database exists but doesn't have any tables in it. I'd really like to get the latest release working with Launchpod so that all the recent fixes can be included in the recommended deploy version. The L deploy worked with the previous version of DMI-TCAT (63080afcf6053fa26204a3302dc5c84eb244330b)

The http error_log has:

```
[Wed Jun 15 01:34:39 2016] [error] [client 131.181.149.145] PHP Fatal error:
Uncaught exception 'PDOException' with message
'SQLSTATE[HY000]: General error: 2014 Cannot execute queries while other
unbuffered queries are active.  Consider using PDOStatement::fetchAll().
Alternatively, if your code is only ever going to run against mysql,
you may enable query buffering by setting the
PDO::MYSQL_ATTR_USE_BUFFERED_QUERY attribute.' in
/var/www/html/dmi-tcat/capture/common/functions.php:310
Stack trace:
#0 /var/www/html/dmi-tcat/capture/common/functions.php(310): PDOStatement->execute()
#1 /var/www/html/dmi-tcat/capture/index.php(13): create_admin()
#2 {main}
  thrown in /var/www/html/dmi-tcat/capture/common/functions.php on line 310
```

Running from the command line results in this error:

```
php dmitcat_track.php PHP Deprecated:  mysql_connect(): The mysql extension
is deprecated and will be removed in the future: use mysqli or PDO instead in
/var/www/html/dmi-tcat/common/functions.php on line 5 PHP Fatal error:  Uncaught
exception 'PDOException' with message 'SQLSTATE[HY000]: General error: 2014
Cannot execute queries while other unbuffered queries are active.  Consider
using PDOStatement::fetchAll(). Alternatively, if your code is only ever going
to run against mysql, you may enable query buffering by setting the
PDO::MYSQL_ATTR_USE_BUFFERED_QUERY attribute.' in
/var/www/html/dmi-tcat/capture/common/functions.php:51
Stack trace:
#0 /var/www/html/dmi-tcat/capture/common/functions.php(51): PDOStatement->execute()
#1 /var/www/html/dmi-tcat/capture/common/functions.php(2028): create_error_logs()
#2 /var/www/html/dmi-tcat/capture/stream/dmitcat_track.php(39): tracker_run()
#3 {main}
thrown in /var/www/html/dmi-tcat/capture/common/functions.php on line 51
```

I tried a second launchpod deploy and the same thing happened.

I tried a deploy on Ubuntu 14.04 using the TCAT install script and it worked. So the problem is that one of the TCAT changes has stopped it working on CentOS 6 and/or with the previous install steps. Do you have any idea what could be causing the problem?

**Comment by dentoir** [comment]
Hi @brendam 

I did some quick test on Amazon, with the following installation:

```
CentOS release 6.7 (Final)
MySQL community server 5.6
Default PHP version from OS: 5.3
Latest dmi-tcat git master version

```

In this configuration the tables are properly set-up, so I can't reproduce the error. However, perhaps you could try the following fix. Edit capture/common/functions.php and insert the following codeblock on line 14, just after the PDO connection has been created:

```
    if (defined('PDO::MYSQL_ATTR_USE_BUFFERED_QUERY')) {
        // Some PHP/PDO builds may sey the MySQL connection by default to use unbuffered queries?
        // This could go awry if we have multiple queries running in parallel, which is simply unsupported.
        $dbh->setAttribute(PDO::MYSQL_ATTR_USE_BUFFERED_QUERY, true);
    }
```

Hope that works.

Cheers,

Emile

**Comment by brendam** [comment]
The server details are:

```
 NeCTAR CentOS release 6.7 (Final) x86_64
mysql  Ver 14.14 Distrib 5.7.13, for Linux (x86_64) using  EditLine wrapper
PHP 5.5.36 (cli) (built: May 25 2016 16:10:36)
Latest dmi-tcat git master version
```


I've tried adding the lines you suggested:
```
     12     $dbh = new PDO("mysql:host=$hostname;dbname=$database;charset=utf8mb4", $dbuser, $dbpass, array(PDO::MYSQL_ATTR_INIT_COMMAND => "set sql_mode='ALLOW_INVALID_D     12 ATES';set time_zone='+00:00]"));
     14     $dbh->setAttribute(PDO::ATTR_ERRMODE, PDO::ERRMODE_EXCEPTION);
     15     if (defined('PDO::MYSQL_ATTR_USE_BUFFERED_QUERY')) {
     16         // Some PHP/PDO builds may sey the MySQL connection by default to use unbuffered queries?
     17         // This could go awry if we have multiple queries running in parallel, which is simply unsupported.
     18         $dbh->setAttribute(PDO::MYSQL_ATTR_USE_BUFFERED_QUERY, true);
     19     }
```


and I still get

```
php dmitcat_track.php
PHP Deprecated:  mysql_connect(): The mysql extension is deprecated and will be removed in the future: use mysqli or PDO instead in /var/www/html/dmi-tcat/common/functions.php on line 5
PHP Fatal error:  Uncaught exception 'PDOException' with message 'SQLSTATE[HY000]: General error: 2014 Cannot execute queries while other unbuffered queries are active.  Consider using PDOStatement::fetchAll().  Alternatively, if your code is only ever going to run against mysql, you may enable query buffering by setting the PDO::MYSQL_ATTR_USE_BUFFERED_QUERY attribute.' in /var/www/html/dmi-tcat/capture/common/functions.php:56
Stack trace:
#0 /var/www/html/dmi-tcat/capture/common/functions.php(56): PDOStatement->execute()
#1 /var/www/html/dmi-tcat/capture/common/functions.php(2033): create_error_logs()
#2 /var/www/html/dmi-tcat/capture/stream/dmitcat_track.php(39): tracker_run()
#3 {main}
  thrown in /var/www/html/dmi-tcat/capture/common/functions.php on line 56
```

and 

```
[Wed Jun 15 10:29:38 2016] [error] [client 118.208.67.254] PHP Fatal error:  Uncaught exception 'PDOException' with message 'SQLSTATE[HY000]: General error: 2014 Cannot execute queries while other unbuffered queries are active.  Consider using PDOStatement::fetchAll().  Alternatively, if your code is only ever going to run against mysql, you may enable query buffering by setting the PDO::MYSQL_ATTR_USE_BUFFERED_QUERY attribute.' in /var/www/html/dmi-tcat/capture/common/functions.php:315\nStack trace:\n#0 /var/www/html/dmi-tcat/capture/common/functions.php(315): PDOStatement->execute()\n#1 /var/www/html/dmi-tcat/capture/index.php(13): create_admin()\n#2 {main}\n  thrown in /var/www/html/dmi-tcat/capture/common/functions.php on line 315
```

**Comment by brendam** [comment]
The Launchpod install is using its own install script, so has anything changed in the way that the latest DMI-TCAT needs to be installed that could be stopping the tables being created?

**Comment by dentoir** [comment]
There are no mandatory new installation steps. As discussed in the wiki, [importing MySQL time zones](https://github.com/digitalmethodsinitiative/dmi-tcat/wiki/Installation-Guide#database-setup) is now advised, but not required.

As a temporary test (_not_ to be used in production), could you see if the installation works if you remove the following part from line 12:

`set time_zone='+00:00';`

I'd have to look at the previous diff if we are 
There are no mandatory new installation steps. As discussed in the wiki, [importing MySQL time zones](https://github.com/digitalmethodsinitiative/dmi-tcat/wiki/Installation-Guide#database-setup) is now advised, but not required.

As a temporary test (_not_ to be used as a fix in production), could you test if the installation works if you remove the following part from line 12:

`set time_zone='+00:00';`

(so basically revert that line to master)

It would surprise me, actually. Your examples involve running dmitcat_track.php from the command-line. I assume you've tested visiting the capture/ interface with your webbrowser. Do you get an identical error?

**Comment by brendam** [comment]
I'll try removing the time zone bit now.
I tried both from the command line and through the webpage - the second block of error text is from http_error.log.

**Comment by brendam** [comment]
Removing `set time_zone='+00:00';` worked, page loaded properly.

**Comment by brendam** [comment]
I'll drop the tables and try again with the `'PDO::MYSQL_ATTR_USE_BUFFERED_QUERY'` 4 lines removed.

**Comment by brendam** [comment]
That worked too, so just removing the `set time_zone='+00:00';` fixes the database creation problem on Launchpod installs.

**Comment by brendam** [comment]
The server is receiving tweets via track, so it seems like the table creation is the only problem.

**Comment by brendam** [comment]
Yes that works!

**Comment by dentoir** [comment]
Perfect, this has now been changed in master.

**Comment by ErikBorra** [comment]
@brendam do you maybe have any suggestions on how we can streamline the addition of new features or code changes w.r.t. the deployment on launchpod?

**Comment by brendam** [comment]
It is late here and I had not saved the updated function.php file before reloading the page. I think there is a syntax error in the code you suggested, but having the timezone line after the error line does work. You were missing the `;` after `INVALID_DATES'`

```
    $dbh = new PDO("mysql:host=$hostname;dbname=$database;charset=utf8mb4", $dbuser, $dbpass, array(PDO::MYSQL_ATTR_INIT_COMMAND => "set sql_mode='ALLOW_INVALID_DATES';"));
    $dbh->setAttribute(PDO::ATTR_ERRMODE, PDO::ERRMODE_EXCEPTION);
    $dbh->query("set time_zone='+00:00'");
```
I think you need to fix that in master, sorry!

**Comment by dentoir** [comment]
I don't think so. There wasn't a ; there before set time_zone was introduced. It's not neccessary if there isn't a second command on the same line. As far as I can see, there is no syntax error in master. Browsing back your comments, it looks like you may have made a edit/save error earlier because you mention your code being this?

`"set sql_mode='ALLOW_INVALID_D     12 ATES';set time_zone='+00:00]"`

**Comment by brendam** [comment]
That was just a line number from displaying the code in `less`. I'll test some more tomorrow.

**Comment by brendam** [comment]
I've just tried a new deploy using Launchpod and it has worked correctly. The only change I made was to set 'utc' in the config.php because they haven't updated to doing that yet. Thanks for fixing this so quickly!

**Comment by brendam** [comment]
@ErikBorra I don't really have any ideas about how to make sure deploys on Launchpod keep working. It will have a 'recommended' DMI-TCAT version set (which will be the current version or newer if one is released before we finish testing). I guess we'll probably have to wait for people to report problems and go from there because testing it for every release may not be realistic? I will see if we can resource having someone one test it regularly if there has been a new release of DMI-TCAT.

**Comment by ErikBorra** [comment]
Hi @brendam,

usually the changes pushed to master are rather small and non-blocking or intrusive. The rate-limit and timezone issues were a reason for bigger code changes. I think that in the first case extra testing is not mandatory. In the second case it seems wiser if someone at your side can verify that it works well on CentOS. Note that we have started a [change log](https://github.com/digitalmethodsinitiative/dmi-tcat/wiki/Change-log), in which we will summarize the changes per commit so that one can get an idea of the impact of certain code changes.

Best,

Erik

(btw, closing this issue as the original problem has been solved and we're starting to get some 'issue drift' here ;))
